[WebAuthn] Implement required CTAP code for SetPinRequest

abigailfox opened the pull request:
#### 66b962b5ff82157ee64fc493c0dfb2d340c3bda2
<pre>
[WebAuthn] Implement required CTAP code for SetPinRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=268871">https://bugs.webkit.org/show_bug.cgi?id=268871</a>
<a href="https://rdar.apple.com/113573055">rdar://113573055</a>

Reviewed by Pascoe.

Implement the required CTAP code to create a SetPinRequest and associated test

* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::SetPinRequest::SetPinRequest):
(fido::pin::SetPinRequest::sharedKey const):
(fido::pin::SetPinRequest::tryCreate):
(fido::pin::encodeAsCBOR):
* Source/WebCore/Modules/webauthn/fido/Pin.h:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/274385@main">https://commits.webkit.org/274385@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/48af982992608ed75b8f99087cc9ef771519fe49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38939 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17870 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41287 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41471 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20736 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15219 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39512 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33778 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13067 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34701 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42748 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35341 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11357 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8717 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -280,6 +280,64 @@ const CryptoKeyAES& TokenRequest::sharedKey() const
+    auto coseKey = encodeCOSEPublicKey(rawPublicKeyResult.returnValue());
+
+    Vector<uint8_t> paddedPin;
+    paddedPin.fill('\0', 64);
```
I think you can do something like this here
```
const size_t minPaddedPinLength = 64;
...
Vector<uint8_t> paddedPin;
paddedPin.reserveInitialCapacity(minPaddedPinLength);
paddedPin.append(inputPin.utf8()); // this line might need to something more
paddedPin.fill('\0', minPaddedPinLength);
```

pascoej commented on Source/WebCore/Modules/webauthn/fido/Pin.h:
```diff
@@ -131,6 +132,22 @@ struct KeyAgreementResponse {
 
+struct SetPinRequest {
+public:
+    WEBCORE_EXPORT const WebCore::CryptoKeyAES& sharedKey() const;
```
This isn't used. We probably don't need to expose it.

pascoej commented on Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
```diff
@@ -183,6 +263,7 @@ TEST(CtapPinTest, TestTokenRequest)
     auto token = TokenRequest::tryCreate(pin, downcast<CryptoKeyEC>(*keyPair.publicKey));
     EXPECT_TRUE(token);
     auto result = encodeAsCBOR(*token);
+
```
erroneous newline

pascoej commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -291,6 +352,15 @@ Vector<uint8_t> encodeAsCBOR(const TokenRequest& request)
 
+Vector<uint8_t> encodeAsCBOR(const SetPinRequest& request)
+{
+    return encodePinCommand(Subcommand::kSetPin, [coseKey = WTFMove(request.m_coseKey), encryptedPin = request.m_newPinEnc, pinUvAuthParam = request.m_pinUvAuthParam] (CBORValue::MapValue* map) mutable {
```
I think we should either unmark `request.m_coseKey` as mutable and not move all the members here (can capture by reference instead) or unmark this `encodeAsCBOR` as const, take in an rvalue, and move all the members.

pascoej commented on Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
```diff
@@ -71,6 +71,86 @@ TEST(CtapPinTest, TestValidateAndConvertToUTF8)
+    EXPECT_FALSE(newPinEncResult.hasException());
+    auto newPinEnc = newPinEncResult.releaseReturnValue();
+
+    const uint8_t expectedNewPinEnc[] = { 0x31, 0x32, 0x33, 0x34, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0 ...
```
We could construct a vector from this and directly compare them to avoid the memcmp below, but I don't feel strongly about it since we memcmp elsewhere in this file.

pascoej commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -280,6 +280,64 @@ const CryptoKeyAES& TokenRequest::sharedKey() const
+    Vector<uint8_t> paddedPin;
+    paddedPin.reserveInitialCapacity(minPaddedPinLength);
+    paddedPin.append(inputPin.utf8().bytes());
+    for (int i = paddedPin.size(); i < 64; i++)
```
`i < minPaddedPinLength`

pascoej commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -280,6 +280,64 @@ const CryptoKeyAES& TokenRequest::sharedKey() const
+        paddedPin.append('\0');
+
+    auto sharedKeySize = sharedKeyHash.size();
+    auto hmacKey = CryptoKeyHMAC::importRaw(sharedKeySize * 8 /* lengthInBits */, CryptoAlgorithmIdentifier::SHA_256, WTFMove(sharedKeyHash), true, CryptoKeyUsageSign);
```
```
auto sharedKeySizeBits = sharedKeyHash.size() * 8;
auto hmacKey = CryptoKeyHMAC::importRaw(sharedKeySizeBits,...
```

johnwilander reviewed (commented):
Nice to see this! It looks like there's only a positive test case. True? If so, please add negative test cases that hit the nullopt early returns.

abigailfox committed:
[WebAuthn] Implement required CTAP code for SetPinRequest
https://bugs.webkit.org/show_bug.cgi?id=268871
rdar://113573055

Reviewed by Pascoe.

Implement the required CTAP code to create a SetPinRequest and associated test

* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
(fido::pin::SetPinRequest::SetPinRequest):
(fido::pin::SetPinRequest::sharedKey const):
(fido::pin::SetPinRequest::tryCreate):
(fido::pin::encodeAsCBOR):
* Source/WebCore/Modules/webauthn/fido/Pin.h:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/274385@main
